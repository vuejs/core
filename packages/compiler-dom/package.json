{
  "name": "@vue/compiler-dom",
  "version": "3.5.13",
  "description": "@vue/compiler-dom",
  "main": "index.js",
  "module": "dist/compiler-dom.esm-bundler.js",
  "types": "dist/compiler-dom.d.ts",
  "unpkg": "dist/compiler-dom.global.js",
  "jsdelivr": "dist/compiler-dom.global.js",
  "files": [
    "index.js",
    "dist"
  ],
  "exports": {
    ".": {
      "types": "./dist/compiler-dom.d.ts",
      "node": {
        "production": "./dist/compiler-dom.cjs.prod.js",
        "development": "./dist/compiler-dom.cjs.js",
        "default": "./index.js"
      },
      "module": "./dist/compiler-dom.esm-bundler.js",
      "import": "./dist/compiler-dom.esm-bundler.js",
      "require": "./index.js"
    },
    "./*": "./*"
  },
  "sideEffects": false,
  "buildOptions": {
    "name": "VueCompilerDOM",
    "compat": true,
    "formats": [
      "esm-bundler",
      "esm-browser",
      "cjs",
      "global"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vuejs/core.git",
    "directory": "packages/compiler-dom"
  },
  "keywords": [
    "vue"
  ],
  "author": "Evan You",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vuejs/core/issues"
  },
  "homepage": "https://github.com/vuejs/core/tree/main/packages/compiler-dom#readme",
  "dependencies": {
    "@vue/shared": "workspace:*",
    "@vue/compiler-core": "workspace:*"
  }
}
