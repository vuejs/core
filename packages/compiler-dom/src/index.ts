import {
  type CodegenResult,
  type CompilerOptions,
  type DirectiveTransform,
  type NodeTransform,
  type ParserOptions,
  type RootNode,
  baseCompile,
  baseParse,
  noopDirectiveTransform,
} from '@vue/compiler-core'
import { parserOptions } from './parserOptions'
import { transformStyle } from './transforms/transformStyle'
import { transformVHtml } from './transforms/vHtml'
import { transformVText } from './transforms/vText'
import { transformModel } from './transforms/vModel'
import { transformOn } from './transforms/vOn'
import { transformShow } from './transforms/vShow'
import { transformTransition } from './transforms/Transition'
import { stringifyStatic } from './transforms/stringifyStatic'
import { ignoreSideEffectTags } from './transforms/ignoreSideEffectTags'
import { extend } from '@vue/shared'

export { parserOptions }

export const DOMNodeTransforms: NodeTransform[] = [
  transformStyle,
  ...(__DEV__ ? [transformTransition] : []),
]

export const DOMDirectiveTransforms: Record<string, DirectiveTransform> = {
  cloak: noopDirectiveTransform,
  html: transformVHtml,
  text: transformVText,
  model: transformModel, // override compiler-core
  on: transformOn, // override compiler-core
  show: transformShow,
}

export function compile(
  src: string | RootNode,
  options: CompilerOptions = {},
): CodegenResult {
  return baseCompile(
    src,
    extend({}, parserOptions, options, {
      nodeTransforms: [
        // ignore <script> and <tag>
        // this is not put inside DOMNodeTransforms because that list is used
        // by compiler-ssr to generate vnode fallback branches
        ignoreSideEffectTags,
        ...DOMNodeTransforms,
        ...(options.nodeTransforms || []),
      ],
      directiveTransforms: extend(
        {},
        DOMDirectiveTransforms,
        options.directiveTransforms || {},
      ),
      transformHoist: __BROWSER__ ? null : stringifyStatic,
    }),
  )
}

export function parse(template: string, options: ParserOptions = {}): RootNode {
  return baseParse(template, extend({}, parserOptions, options))
}

export * from './runtimeHelpers'
export { transformStyle } from './transforms/transformStyle'
export {
  createDOMCompilerError,
  DOMErrorCodes,
  DOMErrorMessages,
} from './errors'
export * from '@vue/compiler-core'
