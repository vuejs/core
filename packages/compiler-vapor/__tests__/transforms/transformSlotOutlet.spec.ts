import { ErrorCodes, NodeTypes } from '@vue/compiler-dom'
import {
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformSlotOutlet,
  transformText,
  transformVBind,
  transformVOn,
  transformVShow,
} from '../../src'
import { makeCompile } from './_utils'

const compileWithSlotsOutlet = makeCompile({
  nodeTransforms: [
    transformText,
    transformSlotOutlet,
    transformElement,
    transformChildren,
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
    show: transformVShow,
  },
})

describe('compiler: transform <slot> outlets', () => {
  test('default slot outlet', () => {
    const { ir, code, helpers } = compileWithSlotsOutlet(`<slot />`)
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('createSlot')
    expect(ir.block.effect).toEqual([])
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'default',
        isStatic: true,
      },
      props: [],
      fallback: undefined,
    })
  })

  test('statically named slot outlet', () => {
    const { ir, code } = compileWithSlotsOutlet(`<slot name="foo" />`)
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'foo',
        isStatic: true,
      },
    })
  })

  test('dynamically named slot outlet', () => {
    const { ir, code } = compileWithSlotsOutlet(`<slot :name="foo + bar" />`)
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'foo + bar',
        isStatic: false,
      },
    })
  })

  test('dynamically named slot outlet with v-bind shorthand', () => {
    const { ir, code } = compileWithSlotsOutlet(`<slot :name />`)
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'name',
        isStatic: false,
      },
    })
  })

  test('default slot outlet with props', () => {
    const { ir, code } = compileWithSlotsOutlet(
      `<slot foo="bar" :baz="qux" :foo-bar="foo-bar" />`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      name: { content: 'default' },
      props: [
        [
          { key: { content: 'foo' }, values: [{ content: 'bar' }] },
          { key: { content: 'baz' }, values: [{ content: 'qux' }] },
          { key: { content: 'fooBar' }, values: [{ content: 'foo-bar' }] },
        ],
      ],
    })
  })

  test('statically named slot outlet with props', () => {
    const { ir, code } = compileWithSlotsOutlet(
      `<slot name="foo" foo="bar" :baz="qux" />`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      name: { content: 'foo' },
      props: [
        [
          { key: { content: 'foo' }, values: [{ content: 'bar' }] },
          { key: { content: 'baz' }, values: [{ content: 'qux' }] },
        ],
      ],
    })
  })

  test('statically named slot outlet with v-bind="obj"', () => {
    const { ir, code } = compileWithSlotsOutlet(
      `<slot name="foo" foo="bar" v-bind="obj" :baz="qux" />`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      name: { content: 'foo' },
      props: [
        [{ key: { content: 'foo' }, values: [{ content: 'bar' }] }],
        { value: { content: 'obj', isStatic: false } },
        [{ key: { content: 'baz' }, values: [{ content: 'qux' }] }],
      ],
    })
  })

  test('statically named slot outlet with v-on', () => {
    const { ir, code } = compileWithSlotsOutlet(
      `<slot @click="foo" v-on="bar" :baz="qux" />`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      props: [
        [{ key: { content: 'click' }, values: [{ content: 'foo' }] }],
        { value: { content: 'bar' }, handler: true },
        [{ key: { content: 'baz' }, values: [{ content: 'qux' }] }],
      ],
    })
  })

  test('default slot outlet with fallback', () => {
    const { ir, code } = compileWithSlotsOutlet(`<slot><div/></slot>`)
    expect(code).toMatchSnapshot()
    expect(ir.template[0]).toBe('<div></div>')
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'default' },
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }],
        },
        returns: [2],
      },
    })
  })

  test('named slot outlet with fallback', () => {
    const { ir, code } = compileWithSlotsOutlet(
      `<slot name="foo"><div/></slot>`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.template[0]).toBe('<div></div>')
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'foo' },
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }],
        },
        returns: [2],
      },
    })
  })

  test('default slot outlet with props & fallback', () => {
    const { ir, code } = compileWithSlotsOutlet(
      `<slot :foo="bar"><div/></slot>`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.template[0]).toBe('<div></div>')
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'default' },
      props: [[{ key: { content: 'foo' }, values: [{ content: 'bar' }] }]],
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }],
        },
        returns: [2],
      },
    })
  })

  test('named slot outlet with props & fallback', () => {
    const { ir, code } = compileWithSlotsOutlet(
      `<slot name="foo" :foo="bar"><div/></slot>`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.template[0]).toBe('<div></div>')
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'foo' },
      props: [[{ key: { content: 'foo' }, values: [{ content: 'bar' }] }]],
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }],
        },
        returns: [2],
      },
    })
  })

  test('error on unexpected custom directive on <slot>', () => {
    const onError = vi.fn()
    const source = `<slot v-foo />`
    const index = source.indexOf('v-foo')
    const { code } = compileWithSlotsOutlet(source, { onError })
    expect(code).toMatchSnapshot()
    expect(onError.mock.calls[0][0]).toMatchObject({
      code: ErrorCodes.X_V_SLOT_UNEXPECTED_DIRECTIVE_ON_SLOT_OUTLET,
      loc: {
        start: {
          offset: index,
          line: 1,
          column: index + 1,
        },
        end: {
          offset: index + 5,
          line: 1,
          column: index + 6,
        },
      },
    })
  })

  test('error on unexpected custom directive with v-show on <slot>', () => {
    const onError = vi.fn()
    const source = `<slot v-show="ok" />`
    const index = source.indexOf('v-show="ok"')
    const { code } = compileWithSlotsOutlet(source, { onError })
    expect(code).toMatchSnapshot()
    expect(onError.mock.calls[0][0]).toMatchObject({
      code: ErrorCodes.X_V_SLOT_UNEXPECTED_DIRECTIVE_ON_SLOT_OUTLET,
      loc: {
        start: {
          offset: index,
          line: 1,
          column: index + 1,
        },
        end: {
          offset: index + 11,
          line: 1,
          column: index + 12,
        },
      },
    })
  })
})
