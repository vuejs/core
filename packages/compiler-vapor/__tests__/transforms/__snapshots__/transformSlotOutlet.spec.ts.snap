// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: transform <slot> outlets > default slot outlet 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot("default", null)
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > default slot outlet with fallback 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue';
const t0 = _template("<div></div>", true)

export function render(_ctx) {
  const n0 = _createSlot("default", null, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > default slot outlet with props & fallback 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue';
const t0 = _template("<div></div>", true)

export function render(_ctx) {
  const n0 = _createSlot("default", { foo: () => (_ctx.bar) }, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > default slot outlet with props 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot("default", {
    foo: () => ("bar"), 
    baz: () => (_ctx.qux), 
    fooBar: () => (_ctx.foo-_ctx.bar)
  })
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > dynamically named slot outlet 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot(() => (_ctx.foo + _ctx.bar), null)
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > dynamically named slot outlet with v-bind shorthand 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot(() => (_ctx.name), null)
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > error on unexpected custom directive on <slot> 1`] = `
"import { resolveDirective as _resolveDirective, createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const _directive_foo = _resolveDirective("foo")
  const n0 = _createSlot("default", null)
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > error on unexpected custom directive with v-show on <slot> 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot("default", null)
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > named slot outlet with fallback 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue';
const t0 = _template("<div></div>", true)

export function render(_ctx) {
  const n0 = _createSlot("foo", null, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > named slot outlet with props & fallback 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue';
const t0 = _template("<div></div>", true)

export function render(_ctx) {
  const n0 = _createSlot("foo", { foo: () => (_ctx.bar) }, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > statically named slot outlet 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot("foo", null)
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > statically named slot outlet with props 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot("foo", {
    foo: () => ("bar"), 
    baz: () => (_ctx.qux)
  })
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > statically named slot outlet with v-bind="obj" 1`] = `
"import { createSlot as _createSlot } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot("foo", {
    foo: () => ("bar"), 
    $: [
      () => (_ctx.obj), 
      { baz: () => (_ctx.qux) }
    ]
  })
  return n0
}"
`;

exports[`compiler: transform <slot> outlets > statically named slot outlet with v-on 1`] = `
"import { createSlot as _createSlot, toHandlers as _toHandlers } from 'vue';

export function render(_ctx) {
  const n0 = _createSlot("default", {
    onClick: () => _ctx.foo, 
    $: [
      () => (_toHandlers(_ctx.bar)), 
      { baz: () => (_ctx.qux) }
    ]
  })
  return n0
}"
`;
